modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/components/filters/SelectionFilter.js

```diff
@@ -113,52 +117,80 @@ function SelectionFilter({
 	const [infiniteLoaderRendered, setInfiniteLoaderRendered] = useState(false);
 	const [exclude, setExclude] = useState(!!selectedData?.exclude);
 
-	const loaderVisible = items?.length < total;
+	const loaderVisible = !localItems.length && items?.length < total;
 
 	useEffect(() => {
 		setSelectedItems(selectedData?.selectedItems || []);
 	}, [selectedData]);
 
-	useEffect(() => {
-		if (query === search) {
-			return;
-		}
-
-		setCurrentPage(1);
-
-		setSearch(query);
-	}, [query, search]);
-
-	const isMounted = useIsMounted();
-
-	useEffect(() => {
-		if (apiURL) {
+	const loadSelectionItems = function () {
+		if (apiURL && !localItems.length) {
 			setLoading(true);
 
 			fetchData(apiURL, search, currentPage)
-				.then((data) => {
+				.then((response) => {
+					const selectionItems = response.items.map((item) => {
+						return {
+							label: itemLabel
+								? getValueFromItem(item, itemLabel)
+								: item.label,
+							value: itemKey ? item[itemKey] : item.value,
+						};
+					});
+
 					if (!isMounted()) {
 						return;
 					}
 
 					setLoading(false);
 
 					if (currentPage === 1) {
-						setItems(data.items);
+						setItems(selectionItems);
 					}
 					else {
-						setItems((items) => [...items, ...data.items]);
+						setItems((items) => [...items, ...selectionItems]);
+					}
+
+					if (
+						firstRequest &&
+						response.totalCount <= DEFAULT_PAGE_SIZE &&
+						autocompleteEnabled
+					) {
+						setLocalItems(selectionItems);
 					}
 
-					setTotal(data.totalCount);
+					setTotal(response.totalCount);
+					setFirstRequest(false);
 				})
 				.catch(() => {
 					if (isMounted()) {
 						setLoading(false);
 					}
 				});
 		}
-	}, [autocompleteEnabled, currentPage, isMounted, search, apiURL]);
+		else if (localItems.length && autocompleteEnabled) {
+			setItems(
+				search
+					? localItems.filter(({label}) =>
+							label.toLowerCase().match(search.toLowerCase())
+					  )
+					: localItems
+			);
+		}
+	};
+
+	const handleAutocompleteQuery = debounce((value) => {
+		setCurrentPage(1);
+		setSearch(value);
+	}, DEFAULT_DEBOUNCE_DELAY);
+
+	const isMounted = useIsMounted();
+
+	useEffect(() => {
+		loadSelectionItems();
+
+		// eslint-disable-next-line react-hooks/exhaustive-deps
```

Comment:
Both `search` and `currentPage` are state variables, so we should be able to refactor code to remove `useEffect`. Conversely, if they were props, `useEffect` might be needed.

For example, there is just one `setSearch` in code
```js
const handleAutocompleteQuery = debounce((value) => {
	setCurrentPage(1);
	setSearch(value);
}, DEFAULT_DEBOUNCE_DELAY);
```

This can be something like 
```js
const handleAutocompleteQuery = debounce((value) => {
	setCurrentPage(1);
	setSearch(value);

	loadSelectionItems({currentPage: 1, search: value});
}, DEFAULT_DEBOUNCE_DELAY);
```
Would that work?

Reply:
We still need to load data the first time the component renders.
Also, if we follow the same approach with the infinite scroll, running the `loadSelectionItems` when the component requests the following pages, we will end up with a similar dependency problem for the [`useCallback` hook](https://github.com/liferay-frontend/liferay-portal/pull/3443/files#diff-fe3d4fd426b630b155b3b84c1e6770b281420e40ff6769549768e2312867e9fdR227)

Reply:
> We still need to load data the first time the component renders.

Ok, for this case we need to leave in `useEffect`, one with an empty array of dependencies. But it's still worth deducing it to just this case.

>Also, if we follow the same approach with the infinite scroll, running the loadSelectionItems when the component requests the following pages, we will end up with a similar dependency problem for the setObserver [useCallback hook](https://github.com/liferay-frontend/liferay-portal/pull/3443/files#diff-fe3d4fd426b630b155b3b84c1e6770b281420e40ff6769549768e2312867e9fdR227)

Yep, we can modify that usage as well. Is there some quirky interaction with `IntersectionObserver` that would prevent us from doing so? :thinking: It should be fine.

Reply:
@markocikos I've found a solution to avoid disabling the es-lint rule
Basically, move the `loadSelectionItems` back to the useEffect hook and add all needed dependencies. However, for non-local search, there is an extra request to the server do to a change in the `firstRequest` state variable.

I could use a React ref to check this variable change, but it only makes this piece of code harder to understand :disappointed: 

```
	const firstRequestRef = useRef(true);

	useEffect(() => {
		if (firstRequestRef.current !== firstRequest) {
			firstRequestRef.current = firstRequest;

			return;
		}

		if (apiURL && !localItems.length) {
               ...
```

Reply:
Can you use only `firstRequestRef` _instead_ of `firstRequest` state variable? `firstRequestRef.current = false` will not cause a re-rendere.